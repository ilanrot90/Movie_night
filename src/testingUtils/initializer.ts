import axiosInstance from 'axiosRequest';
import MocksManager from './mocks/axiosMock';
import { setupIntersectionObserverMock, restoreIntersectionObserverMock } from './mocks/IntersectionObserver';

const originalError = console.error;

beforeAll(() => {
	setupIntersectionObserverMock();
	MocksManager.initialize(axiosInstance);
	console.error = (...args: Array<any>) => {
		if (/Warning.*not wrapped in act/.test(args[0])) {
			return;
		}
		originalError.call(console, ...args);
	};
});

beforeEach(() => {
	MocksManager.reset();
});

afterEach(async () => {
	await MocksManager.validate();
});

afterAll(() => {
	console.error = originalError;
	restoreIntersectionObserverMock();
});
