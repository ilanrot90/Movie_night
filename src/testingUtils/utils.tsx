import React, { ComponentType, ReactElement } from 'react';
import userEvent, { TargetElement } from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';
import App from 'App';
import ThemeProvider from 'style/ThemeProvider';
import { RecoilRoot } from 'recoil';
import { MemoryRouter } from 'react-router-dom';
import { fromPairs } from 'utils/lodash.utils';

const GETTERS = ['Text', 'Role', 'TestId', 'LabelText', 'AltText', 'PlaceholderText'] as const;
const gettersMap = {
	Text: 'findByText' as 'findByText',
	Role: 'findByRole' as 'findByRole',
	TestId: 'findByTestId' as 'findByTestId',
	LabelText: 'findByLabelText' as 'findByLabelText',
	AltText: 'findByAltText' as 'findByAltText',
	PlaceholderText: 'findByPlaceholderText' as 'findByPlaceholderText',
};
const gettersKeys = {
	Text: 'clickByText' as 'findByText',
	Role: 'clickByRole' as 'findByRole',
	TestId: 'clickByTestId' as 'findByTestId',
	LabelText: 'clickByLabelText' as 'findByLabelText',
	AltText: 'clickByAltText' as 'findByAltText',
	PlaceholderText: 'clickByPlaceholderText' as 'findByPlaceholderText',
};
const clickByFactory = (fn: Function) => async (...args: Array<any>) => {
	const el: TargetElement = await fn(...args);
	el && userEvent.click(el);
	return el;
};
const changeByFactory = (fn: Function) => async (value: string, ...args: Array<any>) => {
	const el: TargetElement = await fn(...args);
	el && (await userEvent.type(el, value));
	return el;
};

const customQueries = {
	...fromPairs(GETTERS.map(v => [gettersKeys[v], clickByFactory(screen[gettersMap[v]])])),
	...fromPairs(GETTERS.map(v => [gettersKeys[v], changeByFactory(screen[gettersMap[v]])])),
};
// helper to test login screens
export const signOutRender = async () => {
	renderUi({ route: '/app' });
	expect(screen.getByRole('button', { name: /log out/i })).toBeInTheDocument();
	userEvent.click(screen.getByRole('button', { name: /log out/i }));
	return await waitFor(() => expect(screen.getAllByRole('button', { name: /log in/i })).toBeArrayOfSize(4));
};

const renderProviders = (route: string) => ({ children }: { children: ReactElement }): ReactElement => (
	<RecoilRoot>
		<ThemeProvider>
			<MemoryRouter initialEntries={[route]}> {children}</MemoryRouter>
		</ThemeProvider>
	</RecoilRoot>
);
type RenderOptions = {
	route: string;
};
export function renderUi({ route, ...renderOptions }: RenderOptions) {
	return render(<App />, { wrapper: renderProviders(route) as ComponentType, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
export { userEvent, customQueries };
