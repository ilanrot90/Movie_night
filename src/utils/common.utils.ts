export const isProduction = process.env.NODE_ENV === 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const asyncHandler = <T>(promise: Promise<T>) => {
	return promise
		.then(response => ({
			error: undefined,
			response,
		}))
		.catch(error => ({
			response: undefined,
			error,
		}));
};
