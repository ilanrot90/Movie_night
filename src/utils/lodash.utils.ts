import keyBy from 'lodash/keyBy';
import mapValues from 'lodash/mapValues';
import map from 'lodash/map';
import orderBy from 'lodash/orderBy';
import groupBy from 'lodash/groupBy';
import isEmpty from 'lodash/isEmpty';

const chainableFunctions = {
	map,
	orderBy,
	groupBy,
};

const chain = (input: any) => {
	let value = input;
	const wrapper = {
		...mapValues(chainableFunctions, f => (...args: any) => {
			// @ts-ignore
			value = f(...args);
			return wrapper;
		}),
		value: () => value,
	};

	return wrapper;
};

export { chain, map, orderBy, groupBy, keyBy, isEmpty };
