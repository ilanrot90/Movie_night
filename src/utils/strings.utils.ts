export const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
export const nameRegex = /^[a-zA-Z\s]*$/;
export const extendHex = (v: string): string => v.replace(/#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])/, '#$1$1$2$2$3$3');
