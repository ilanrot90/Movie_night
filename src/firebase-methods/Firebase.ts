// firebase-methods imports
import app from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';

import { firebaseDevConfig, firebaseProdConfig } from 'config';
import { isProduction } from 'utils/common.utils';
// config
const config = isProduction ? firebaseProdConfig : firebaseDevConfig;
// providers
const googleProvider = new app.auth.GoogleAuthProvider().addScope('https://www.googleapis.com/auth/userinfo.email');
const facebookProvider = new app.auth.FacebookAuthProvider().addScope('email');
const githubProvider = new app.auth.GithubAuthProvider().addScope('user:email');

export type Provider = 'google' | 'facebook' | 'github';
export const getProvider = (provider: Provider) =>
	({
		google: googleProvider,
		facebook: facebookProvider,
		github: githubProvider,
	}[provider]);

export const firebase = app.initializeApp(config);
export const auth = firebase.auth();
// export const emailAuthProvider = app.auth.EmailAuthProvider;
export const firestore = firebase.firestore();
