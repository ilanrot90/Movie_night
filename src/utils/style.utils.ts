import { extendHex } from './strings.utils';

const isHexColor = (v: string): boolean => /^#([A-Fa-f0-9]{3}){1,2}$/.test(v);

export type HexToRgba = (hex: string, a?: number) => string;
export const hexToRgba: HexToRgba = (hex, a = 1) => {
	if (!isHexColor(hex)) {
		throw new Error('Bad Hex');
	}
	const h = hex.length === 4 ? extendHex(hex) : hex;

	const r = parseInt(h.slice(1, 3), 16);
	const g = parseInt(h.slice(3, 5), 16);
	const b = parseInt(h.slice(5, 7), 16);

	return `rgba(${r}, ${g}, ${b}, ${a})`;
};
