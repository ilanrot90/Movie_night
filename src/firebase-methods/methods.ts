import { facebookProvider, githubProvider, googleProvider, auth, firestore } from './Firebase';
import { Provider, FormValues } from 'types';
import { asyncHandler } from 'utils/common.utils';
import { get } from 'utils/lodash.utils';

export const handleEmailSignUp = async ({ email, password, displayName }: FormValues) => {
	const { error } = await asyncHandler(auth.createUserWithEmailAndPassword(email, password));
	if (error) {
		// TODO: handle error
		console.log('handleEmailSignIn', { error });
		return error;
	}
	const user = auth.currentUser;

	// send verify email
	await user?.sendEmailVerification();
	// add to DB
	await firestore.collection('users').doc(user?.uid).set({
		email,
		displayName,
	});

	return user;
};

// Send recover password
export const recoverPassword = async (email: string) => {
	const { error, response } = await asyncHandler(auth.sendPasswordResetEmail(email));
	if (error) {
		// TODO: handle error
		console.log('recoverPassword', { error });
		return error;
	}
	return response;
};
// resend email validation
export const resendVerifyEmail = async () => {
	const user = auth.currentUser;

	if (user) {
		const { error } = await asyncHandler(user.sendEmailVerification());
		if (error) {
			// TODO: handle error
			console.log('resendVerifyEmail', { error });
			return error;
		}
		return user;
	}
	return;
};
export const getProvider = (provider: Provider) =>
	({
		google: googleProvider,
		facebook: facebookProvider,
		github: githubProvider,
	}[provider]);

export const loginWithProvider = async (provider: Provider) => {
	let isNewUser = false;
	const providerConfig = getProvider(provider);
	const { response, error } = await asyncHandler(auth.signInWithPopup(providerConfig));
	if (error) {
		console.log('loginWithProvider', { error });
		throw error;
	}

	if (response?.user) {
		const { uid } = response.user;
		const { email, displayName } = get(response, 'user.providerData[0]');

		if (response?.additionalUserInfo?.isNewUser) {
			await Promise.all([
				auth.currentUser?.updateProfile({
					displayName,
				}),
				firestore.collection('users').doc(uid).set({
					email,
					displayName,
				}),
			]);

			isNewUser = true;
		}

		return { isNewUser, user: get(response, 'user') };
	}

	throw new Error('Internal error');
};

export const handleEmailLogin = async ({ email, password }: FormValues) => {
	const { response, error } = await asyncHandler(auth.signInWithEmailAndPassword(email, password));
	if (error) {
		console.log('handleEmailLogin', { error });
		return error;
	}

	return response?.user;
};
