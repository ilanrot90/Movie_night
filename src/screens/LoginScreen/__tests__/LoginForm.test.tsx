import { signOutRender, userEvent, waitFor, screen } from 'testingUtils/utils';
import { TargetElement } from '@testing-library/user-event';
import { emailNotValidMessage, passwordValidation } from '../authScreens.utils';
import * as firebaseMethods from 'firebase-methods/methods';

const mockedLoginWithProvider = jest.spyOn(firebaseMethods, 'loginWithProvider');
const [validEmail, unValidEmail] = ['test@test.com', 'test-test.com'];
const [validPassword, unValidPassword] = ['min_6_length', '12345'];

describe('Login actions', () => {
	let emailInput: TargetElement;
	let passwordInput: TargetElement;
	let loginButton: TargetElement;

	beforeEach(async () => {
		await signOutRender();
		emailInput = screen.getByRole('textbox', {
			name: /email/i,
		});
		passwordInput = screen.getByPlaceholderText(/password/i);
		loginButton = screen.getByTestId(/password-login/i);
	});

	afterAll(() => {
		mockedLoginWithProvider.mockRestore();
	});

	test('Successful Login with Password', async () => {
		await userEvent.type(emailInput, validEmail);
		await userEvent.type(passwordInput, validPassword);

		expect(emailInput).toHaveValue(validEmail);
		expect(passwordInput).toHaveValue(validPassword);

		await userEvent.click(loginButton);

		await waitFor(() => expect(screen.getByRole('button', { name: /log out/i })).toBeInTheDocument());
	});

	test('Login button is Disabled when login in process', async () => {
		await userEvent.type(emailInput, validEmail);
		await userEvent.type(passwordInput, validPassword);
		await userEvent.click(loginButton);

		await waitFor(() => expect(loginButton).toBeDisabled());
	});

	test('Fields not valid format - show errors', async () => {
		await userEvent.type(emailInput, unValidEmail);
		await userEvent.type(passwordInput, unValidPassword);
		await userEvent.click(loginButton);

		const errorPasswordMsg = await screen.findByText(emailNotValidMessage);
		const errorEmailMsg = await screen.findByText(passwordValidation);

		expect(errorPasswordMsg).toBeInTheDocument();
		expect(errorEmailMsg).toBeInTheDocument();
	});

	test('Sign in with provider', async () => {
		const loginButtons = screen.getAllByRole('button', { name: /log in with/i });
		expect(loginButtons).toMatchSnapshot();

		const facebookLoginBtn = loginButtons[0];
		await userEvent.click(facebookLoginBtn);

		expect(mockedLoginWithProvider).toHaveBeenCalledTimes(1);
		expect(mockedLoginWithProvider).toHaveBeenCalledWith('facebook');
	});

	test('Show error message when login failed', async () => {
		mockedLoginWithProvider.mockReset();
		mockedLoginWithProvider.mockRejectedValue(new Error('Failed login with facebook'));
		const loginButtons = screen.getAllByRole('button', { name: /log in with/i });

		const facebookLoginBtn = loginButtons[0];
		await userEvent.click(facebookLoginBtn);

		expect(mockedLoginWithProvider).toHaveBeenCalledTimes(1);
		expect(mockedLoginWithProvider).toHaveBeenCalledWith('facebook');

		const errorPopup = await screen.findByTestId(/login-screens-error-msg/i);
		expect(errorPopup).toMatchSnapshot();

		expect(errorPopup).toBeInTheDocument();
	});
});
