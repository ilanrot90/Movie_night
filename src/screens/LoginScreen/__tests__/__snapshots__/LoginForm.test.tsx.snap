// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login actions Show error message when login failed 1`] = `
<div
  class="MuiPaper-root MuiAlert-root MuiAlert-filledError MuiPaper-elevation6"
  data-testid="login-screens-error-msg"
  direction="up"
  role="alert"
  style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
>
  <div
    class="MuiAlert-icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message"
  >
    Failed login with facebook
  </div>
  <div
    class="MuiAlert-action"
  >
    <button
      aria-label="Close"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
      tabindex="0"
      title="Close"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Login actions Sign in with provider 1`] = `
Array [
  .c0 {
  height: 40px;
  background-color: #3b5998;
}

.c0:hover {
  background-color: #2b56b1;
  box-shadow: 0 14px 26px -12px rgba(51,51,51,0.42),0 4px 23px 0px rgba(0,0,0,0.12), 0 8px 10px -5px rgba(51,51,51,0.2);
}

.c0.MuiButton-contained.Mui-disabled {
  color: #f9fafa;
  opacity: 0.5;
  background-color: #3f50b5;
}

.c1 {
  margin: 8px 0;
}

.c1 .icon {
  position: absolute;
  top: 10px;
  left: 24px;
}

.c2 {
  fill: #f9fafa;
}

<button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained c0 c1 MuiButton-containedPrimary MuiButton-fullWidth"
    data-testid="btn"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      log in with 
      facebook
      <span
        class="MuiButton-endIcon MuiButton-iconSizeMedium"
      >
        <svg
          class="c2 undefined icon-facebook icon"
          height="22"
          viewBox="0 0 1024 1024"
          width="22"
        >
          <g
            id="facebook"
          >
            <path
              d="M928 0h-832c-52.8 0-96 43.2-96 96v832c0 52.8 43.2 96 96 96h416v-448h-128v-128h128v-64c0-105.8 86.2-192 192-192h128v128h-128c-35.2 0-64 28.8-64 64v64h192l-32 128h-160v448h288c52.8 0 96-43.2 96-96v-832c0-52.8-43.2-96-96-96z"
            />
          </g>
        </svg>
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>,
  .c0 {
  height: 40px;
  background-color: #dd4b39;
}

.c0:hover {
  background-color: #e23824;
  box-shadow: 0 14px 26px -12px rgba(51,51,51,0.42),0 4px 23px 0px rgba(0,0,0,0.12), 0 8px 10px -5px rgba(51,51,51,0.2);
}

.c0.MuiButton-contained.Mui-disabled {
  color: #f9fafa;
  opacity: 0.5;
  background-color: #3f50b5;
}

.c1 {
  margin: 8px 0;
}

.c1 .icon {
  position: absolute;
  top: 10px;
  left: 24px;
}

.c2 {
  fill: #f9fafa;
}

<button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained c0 c1 MuiButton-containedPrimary MuiButton-fullWidth"
    data-testid="btn"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      log in with 
      google
      <span
        class="MuiButton-endIcon MuiButton-iconSizeMedium"
      >
        <svg
          class="c2 undefined icon-google icon"
          height="22"
          viewBox="0 0 1024 1024"
          width="22"
        >
          <g
            id="google"
          >
            <path
              d="M928 0h-832c-52.8 0-96 43.2-96 96v832c0 52.8 43.2 96 96 96h832c52.8 0 96-43.2 96-96v-832c0-52.8-43.2-96-96-96zM519.6 896c-212.2 0-384-171.8-384-384s171.8-384 384-384c103.6 0 190.4 37.8 257.2 100.4l-104.2 100.4c-28.6-27.4-78.4-59.2-153-59.2-131.2 0-238 108.6-238 242.4s107 242.4 238 242.4c152 0 209-109.2 217.8-165.6h-217.8v-131.6h362.6c3.2 19.2 6 38.4 6 63.6 0.2 219.4-146.8 375.2-368.6 375.2z"
            />
          </g>
        </svg>
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>,
  .c0 {
  height: 40px;
  background-color: #333333;
}

.c0:hover {
  background-color: #2d2c2c;
  box-shadow: 0 14px 26px -12px rgba(51,51,51,0.42),0 4px 23px 0px rgba(0,0,0,0.12), 0 8px 10px -5px rgba(51,51,51,0.2);
}

.c0.MuiButton-contained.Mui-disabled {
  color: #f9fafa;
  opacity: 0.5;
  background-color: #3f50b5;
}

.c1 {
  margin: 8px 0;
}

.c1 .icon {
  position: absolute;
  top: 10px;
  left: 24px;
}

.c2 {
  fill: #f9fafa;
}

<button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained c0 c1 MuiButton-containedPrimary MuiButton-fullWidth"
    data-testid="btn"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      log in with 
      github
      <span
        class="MuiButton-endIcon MuiButton-iconSizeMedium"
      >
        <svg
          class="c2 undefined icon-github icon"
          height="22"
          viewBox="0 0 1024 1024"
          width="22"
        >
          <g
            id="github"
          >
            <path
              d="M296.664 685.775v0c1.143-1.715 0.571-5.145-1.715-7.432-2.858-2.287-6.288-2.858-8.003-1.143-1.143 1.715-0.571 5.145 1.715 7.432 2.858 2.287 6.288 2.858 8.003 1.143zM280.659 662.339c-1.715-2.287-4.573-3.43-6.859-2.287-1.715 1.143-1.715 4.573 0 6.859 2.287 2.858 5.145 4.001 6.859 2.858v0c1.715-1.143 1.715-4.573 0-7.432zM257.223 639.474v0c0.571-1.143-0.571-3.43-2.858-4.573-1.715-0.571-4.001-0.571-4.573 1.143-1.143 1.715 0 3.43 2.287 4.573 2.287 0.571 4.573 0.571 5.145-1.143zM269.227 652.62v0c1.143-1.143 1.143-4.001-1.143-5.716-1.715-2.287-4.573-2.858-5.716-1.715-1.715 1.715-1.143 4.001 0.571 6.288 1.715 1.715 4.573 2.858 6.288 1.143zM318.387 695.493v0c0.571-2.287-1.715-5.145-5.145-6.288s-6.288 0-7.432 2.287c-0.571 2.287 1.715 5.145 5.145 6.288s6.288 0 7.432-2.287zM342.395 697.208v0c0-2.287-2.858-4.573-6.859-4.573-3.43 0-5.716 2.287-5.716 4.573s2.858 4.573 6.288 4.573 6.288-2.287 6.288-4.573zM364.687 693.206v0c-0.571-2.287-4.001-3.43-7.432-2.858s-5.716 2.858-5.145 5.145 3.43 4.001 6.859 3.43 5.716-3.43 5.716-5.716zM731.672 512.001c0-161.77-130.902-292.672-292.672-292.672s-292.672 130.902-292.672 292.672c0 129.187 84.029 238.939 200.069 277.809 14.862 2.858 20.007-6.288 20.007-14.291 0-6.859 0-29.724-0.571-54.305 0 0-81.171 17.72-98.32-34.869 0 0-13.148-33.726-32.582-42.301 0 0-26.295-18.292 2.287-18.292 0 0 28.581 2.287 44.586 30.296 25.723 45.159 68.595 32.011 85.172 24.579 2.858-18.863 10.29-32.011 18.863-39.443-65.165-7.432-133.76-32.582-133.76-144.621 0-32.011 11.433-57.734 30.296-78.313-2.858-7.432-13.148-37.156 2.858-77.741 24.579-7.432 80.599 30.296 80.599 30.296 23.437-6.859 48.017-9.717 73.168-9.717s49.731 2.858 73.168 9.717c0 0 56.019-37.727 80.599-30.296 16.006 40.585 5.716 70.31 2.858 77.741 18.863 20.578 30.296 46.302 30.296 78.313 0 112.61-68.595 137.19-133.76 144.621 10.29 9.146 20.007 26.866 20.007 54.305 0 38.87-0.571 70.31-0.571 80.028 0 8.003 5.145 17.149 20.007 14.291 116.040-38.87 200.069-148.622 200.069-277.809zM878.007 237.621v548.76c0 90.888-73.739 164.628-164.628 164.628h-548.76c-90.888 0-164.628-73.739-164.628-164.628v-548.76c0-90.888 73.739-164.628 164.628-164.628h548.76c90.888 0 164.628 73.739 164.628 164.628z"
            />
          </g>
        </svg>
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>,
]
`;
